--- Day 5: If You Give A Seed A Fertilizer ---
You take the boat and find the gardener right where you were told he would be: managing a giant "garden" that looks more to you like a farm.

"A water source? Island Island is the water source!" You point out that Snow Island isn't receiving any water.

"Oh, we had to stop the water because we ran out of sand to filter it with! Can't make snow with dirty water. Don't worry, I'm sure we'll get more sand soon; we only turned off the water a few days... weeks... oh no." His face sinks into a look of horrified realization.

"I've been so busy making sure everyone here has food that I completely forgot to check why we stopped getting more sand! There's a ferry leaving soon that is headed over in that direction - it's much faster than your boat. Could you please go check it out?"

You barely have time to agree to this request when he brings up another. "While you wait for the ferry, maybe you can help us with our food production problem. The latest Island Island Almanac just arrived and we're having trouble making sense of it."

The almanac (your puzzle input) lists all of the seeds that need to be planted. It also lists what type of soil to use with each kind of seed, what type of fertilizer to use with each kind of soil, what type of water to use with each kind of fertilizer, and so on. Every type of seed, soil, fertilizer and so on is identified with a number, but numbers are reused by each category - that is, soil 123 and fertilizer 123 aren't necessarily related to each other.

For example:

seeds: 79 14 55 13

seed-to-soil map:
50 98 2
52 50 48

soil-to-fertilizer map:
0 15 37
37 52 2
39 0 15

fertilizer-to-water map:
49 53 8
0 11 42
42 0 7
57 7 4

water-to-light map:
88 18 7
18 25 70

light-to-temperature map:
45 77 23
81 45 19
68 64 13

temperature-to-humidity map:
0 69 1
1 0 69

humidity-to-location map:
60 56 37
56 93 4
The almanac starts by listing which seeds need to be planted: seeds 79, 14, 55, and 13.

The rest of the almanac contains a list of maps which describe how to convert numbers from a source category into numbers in a destination category. That is, the section that starts with seed-to-soil map: describes how to convert a seed number (the source) to a soil number (the destination). This lets the gardener and his team know which soil to use with which seeds, which water to use with which fertilizer, and so on.

Rather than list every source number and its corresponding destination number one by one, the maps describe entire ranges of numbers that can be converted. Each line within a map contains three numbers: the destination range start, the source range start, and the range length.

Consider again the example seed-to-soil map:

50 98 2
52 50 48
The first line has a destination range start of 50, a source range start of 98, and a range length of 2. This line means that the source range starts at 98 and contains two values: 98 and 99. The destination range is the same length, but it starts at 50, so its two values are 50 and 51. With this information, you know that seed number 98 corresponds to soil number 50 and that seed number 99 corresponds to soil number 51.

The second line means that the source range starts at 50 and contains 48 values: 50, 51, ..., 96, 97. This corresponds to a destination range starting at 52 and also containing 48 values: 52, 53, ..., 98, 99. So, seed number 53 corresponds to soil number 55.

Any source numbers that aren't mapped correspond to the same destination number. So, seed number 10 corresponds to soil number 10.

So, the entire list of seed numbers and their corresponding soil numbers looks like this:

seed  soil
0     0
1     1
...   ...
48    48
49    49
50    52
51    53
...   ...
96    98
97    99
98    50
99    51
With this map, you can look up the soil number required for each initial seed number:

Seed number 79 corresponds to soil number 81.
Seed number 14 corresponds to soil number 14.
Seed number 55 corresponds to soil number 57.
Seed number 13 corresponds to soil number 13.
The gardener and his team want to get started as soon as possible, so they'd like to know the closest location that needs a seed. Using these maps, find the lowest location number that corresponds to any of the initial seeds. To do this, you'll need to convert each seed number through other categories until you can find its corresponding location number. In this example, the corresponding types are:

Seed 79, soil 81, fertilizer 81, water 81, light 74, temperature 78, humidity 78, location 82.
Seed 14, soil 14, fertilizer 53, water 49, light 42, temperature 42, humidity 43, location 43.
Seed 55, soil 57, fertilizer 57, water 53, light 46, temperature 82, humidity 82, location 86.
Seed 13, soil 13, fertilizer 52, water 41, light 34, temperature 34, humidity 35, location 35.
So, the lowest location number in this example is 35.

What is the lowest location number that corresponds to any of the initial seed numbers?

In [20]:
with open('input05.txt') as file:
    data_str = file.read()

In [21]:
blocks = data_str.split(' map:')
blocks

['seeds: 564468486 226119074 3264322485 135071952 3144185610 89794463 1560365167 555667043 2419038624 7808461 1264209624 9380035 105823719 425973940 4115473551 104486997 3784956593 300187503 975280918 257129208\n\nseed-to-soil',
 '\n1383244180 2567207479 366571891\n3099137706 184794382 526583321\n3625721027 2933779370 589006054\n1749816071 838554248 828858952\n4214727081 3750906360 80240215\n2585129273 2301737819 265469660\n919423459 3831146575 463820721\n2850598933 2053199046 248538773\n184794382 1667413200 385785846\n570580228 3522785424 221666686\n2578675023 3744452110 6454250\n792246914 711377703 127176545\n\nsoil-to-fertilizer',
 '\n1263133301 1101112301 247905608\n0 1881569198 406369887\n3891730222 4233430396 61536900\n2723511851 1349017909 413665324\n2115333016 1762683233 118885965\n4272202829 4116902795 22764467\n1668270240 169908659 417260703\n4047030256 3891730222 225172573\n1511038909 990025537 111086764\n3953267122 4139667262 93763134\n2085530943 960223464 29802073\n1622125

In [22]:
import re
seeds = {int(x) : None for x in re.findall(r'\d+',blocks[0])}
seeds

{564468486: None,
 226119074: None,
 3264322485: None,
 135071952: None,
 3144185610: None,
 89794463: None,
 1560365167: None,
 555667043: None,
 2419038624: None,
 7808461: None,
 1264209624: None,
 9380035: None,
 105823719: None,
 425973940: None,
 4115473551: None,
 104486997: None,
 3784956593: None,
 300187503: None,
 975280918: None,
 257129208: None}

In [23]:
map_code = []
for block in blocks[1:]:
    triple_str = re.findall(r'^\d+ \d+ \d+$', block, flags=re.MULTILINE)
    triples = [tuple(int(x) for x in string.split()) for string in triple_str]
    map_code.append(triples)
map_code[0]

[(1383244180, 2567207479, 366571891),
 (3099137706, 184794382, 526583321),
 (3625721027, 2933779370, 589006054),
 (1749816071, 838554248, 828858952),
 (4214727081, 3750906360, 80240215),
 (2585129273, 2301737819, 265469660),
 (919423459, 3831146575, 463820721),
 (2850598933, 2053199046, 248538773),
 (184794382, 1667413200, 385785846),
 (570580228, 3522785424, 221666686),
 (2578675023, 3744452110, 6454250),
 (792246914, 711377703, 127176545)]

In [24]:
def convert(input, layer_map):
    for dest_r, source_r, len_r in layer_map:
        if source_r <= input <= source_r + len_r - 1:
            return dest_r + input - source_r
    return input

In [25]:
for seed in seeds:
    v = seed
    for layer in map_code:
        v = convert(v, layer)
    seeds[seed] = v
seeds

{564468486: 2422809993,
 226119074: 3367027469,
 3264322485: 697158568,
 135071952: 1136789458,
 3144185610: 2725834042,
 89794463: 3527673368,
 1560365167: 1195534841,
 555667043: 2414008550,
 2419038624: 517243940,
 7808461: 3859069353,
 1264209624: 601644709,
 9380035: 3860640927,
 105823719: 2408170138,
 425973940: 1585436691,
 4115473551: 1291781871,
 104486997: 2406833416,
 3784956593: 340994526,
 300187503: 1648034705,
 975280918: 870127997,
 257129208: 1604976410}

In [26]:
min(seeds.values())

340994526

Your puzzle answer was 340994526.

The first half of this puzzle is complete! It provides one gold star: *

--- Part Two ---
Everyone will starve if you only plant such a small number of seeds. Re-reading the almanac, it looks like the seeds: line actually describes ranges of seed numbers.

The values on the initial seeds: line come in pairs. Within each pair, the first value is the start of the range and the second value is the length of the range. So, in the first line of the example above:

seeds: 79 14 55 13
This line describes two ranges of seed numbers to be planted in the garden. The first range starts with seed number 79 and contains 14 values: 79, 80, ..., 91, 92. The second range starts with seed number 55 and contains 13 values: 55, 56, ..., 66, 67.

Now, rather than considering four seed numbers, you need to consider a total of 27 seed numbers.

In the above example, the lowest location number can be obtained from seed number 82, which corresponds to soil 84, fertilizer 84, water 84, light 77, temperature 45, humidity 46, and location 46. So, the lowest location number is 46.

Consider all of the initial seed numbers listed in the ranges on the first line of the almanac. What is the lowest location number that corresponds to any of the initial seed numbers?

In [27]:
# Looking at the puzzle input, the lowest location numbers are given by the humidity to location map 0 149088969 236814486 (the range after that would be 236814486 5872218 143216751)
# Let's see if we can reverse engineer what kind of numbers result in them

In [28]:
for m in map_code:
    m.sort() # sort by destination value

# map with reversed layers and (start idx, offset)-tuples for easier processing
r_map = []
for orig in reversed(map_code):
    layer = {0: 0}
    for dest, source, r_len in orig:
        assert max(layer) == 0 or dest >= max(layer)
        layer[dest] = source - dest
        layer[dest + r_len] = 0
    layer.setdefault(0, 0)
    r_map.append(layer)
r_map

[{0: 149088969,
  236814486: -230942268,
  380031237: -380031237,
  385903455: 0,
  437631438: 1444512720,
  565280258: 177353367,
  828280122: 1776880938,
  1025526567: 2808131420,
  1091954734: 1905220197,
  1110445989: 1486496765,
  1118664295: 310490049,
  1197665086: -557037833,
  1299671458: 2798945392,
  1399091075: 752461955,
  1456030811: 2045388098,
  1517885288: 125395912,
  1719071266: 533121472,
  1753847939: -539807056,
  1765728190: 1036679315,
  1806366562: 2391669905,
  1856685687: -348530552,
  1991811752: 1908274402,
  2035751658: -155380258,
  2037524416: 840777681,
  2156397250: 1479095731,
  2348539881: -504072703,
  2384444103: -982792234,
  2411817723: 431228154,
  2447073943: -1147769586,
  2549421455: 1013851931,
  2621641050: 871898517,
  2629520392: -2191888954,
  2832516207: -545546796,
  3086070204: -1657044715,
  3086199059: -2080565570,
  3268322772: -2080565570,
  3294606453: -754083045,
  3342903047: 601123013,
  3454152286: -1245659520,
  3497852258: 

In [ ]:
def find_ranges(lower, upper, rl_map):
    intermediates = []
    for start, offset in rl_map.items():
        if start <= lower:
            lower_offset = offset
        if lower < start < upper:
            intermediates.append((start, offset))
    populated = [(lower, lower_offset)] + intermediates
    ranges = []
    for (start, offset), (next_start, _) in zip(populated, populated[1:]):
        low = start + offset
        high = next_start - 1 + offset
        ranges.append((low, high))
    # final range:
    start, offset = populated[-1]
    low = start + offset
    high = upper + offset
    ranges.append((low, high))
    return ranges

In [30]:
initial_range = (0,236814485)
ranges = [initial_range]
for layer in r_map:
    new_ranges = []
    for l, h in ranges:
        new_ranges.extend(find_ranges(l, h, layer))
    ranges = new_ranges
seed_ranges = ranges
seed_ranges

[(2141268585, 2193479228),
 (385733775, 395351339),
 (1431743533, 1447779769),
 (1638613288, 1640301588),
 (3805257397, 3828093567),
 (1600710845, 1617685827),
 (2044229553, 2053199045),
 (3522785424, 3539374557),
 (3871946580, 3901748652),
 (851421410, 852210110),
 (3976531589, 4012835416),
 (2318348284, 2343345639)]

In [31]:
# these ranges correspond to the lowest location numbers and are already sorted. Which of these also correspond to a seed?

seeds = [int(x) for x in re.findall(r'\d+',blocks[0])]
seeds = [(x, x+y-1) for x, y in zip(seeds[::2], seeds[1::2])]
seeds

[(564468486, 790587559),
 (3264322485, 3399394436),
 (3144185610, 3233980072),
 (1560365167, 2116032209),
 (2419038624, 2426847084),
 (1264209624, 1273589658),
 (105823719, 531797658),
 (4115473551, 4219960547),
 (3784956593, 4085144095),
 (975280918, 1232410125)]

In [35]:
def find_lowest(seeds, candidates):
    seeds = sorted(seeds)
    candidates = sorted(candidates)
    for low_can, high_can in candidates:
        for low_seed, high_seed in seeds:
            if low_can <= low_seed <= high_can:
                return low_seed
            elif low_seed <= low_can <= high_seed:
                return low_can

find_lowest(seeds, seed_ranges)

385733775

In [36]:
v = find_lowest(seeds, seed_ranges)
for layer in map_code:
    v = convert(v, layer)
v

52210644

That's the right answer! You are one gold star closer to restoring snow operations.